### ISSUE TITLE
[WSL] Mongo fails to start for db's on a shared WSL/windows filesystem

### ISSUE THREAD

**Issue by brucejo75** [meteor/meteor#9914]
This may have worked prior to V1.4, see https://github.com/meteor/meteor-feature-requests/issues/154#issuecomment-316467990 by @mizzao.

### Environment
Windows 10
WSL
Meteor v1.6.1.1

### Repro
Repro a working case:
1. `cd ~`
2. `meteor create wsltest`
3. `cd wsltest; meteor`
4. mongo starts up and **Meteor works great!**

**Note**: the working mongodb command looks like this:
```bash
/home/bruce/.meteor/packages/meteor-tool/.1.6.1_1.q2yqvp.vmez9++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/mongodb/bin/mongod --bind_ip 127.0.0.1
 --port 3001 --dbpath /home/bruce/wsltest/.meteor/local/db --oplogSize 8 --replSet meteor --noauth --nojournal
```

Repro the fail case:
1. go to a mounted drive e.g. (`E:\temp`) `cd /mnt/e/temp`
2. `meteor create wsltest`
3. `cd wsltest; meteor`
4. mongo fails to start up with an error from mongo: `Unexpected mongo exit code 100. Restarting.`

### Observations

This https://github.com/meteor/meteor-feature-requests/issues/154#issuecomment-316467990 makes me think it may be a simple problem.

#### `--dbpath` problem?
I wonder if there is just a simple problem of not setting `--dbpath` correctly?

For the 2nd case the `--dbpath` should be:
`/mnt/e/temp/wsltest/.meteor/local/db`

#### `~` subdirectories are the only place mongodb can be run?
The `~` directory is special within WSL and may support the file capabilities that mongo requires.  For example, windows tools will mess up the filesystem if they write to files in the `~` directory.

**Comment by WanielDeiss** [comment]
Hey brucejo75,

i got the same problem, my solution is a symlink to the linux subsystem. It's not the best solution for this, but it works.

The path on my Windows mashine is:
`c:/Users/leem/Documents/Projects/bossplate`
On Linux Subsystem:
`/mnt/c/Users/leemb/Documents/Projects/bossplate`

Now i create on my Linux Subsystem User a Folder: 
`~/projects/bossplate-meteor`

Then create a Symlink to this folder on the mounted Windows folder.
```
cd /mnt/c/Users/leemb/Documents/Projects/bossplate/.meteor
rm -rf local
ln -s ~/projects/bossplate-meteor local
```

After this, my Meteor project works fine.

**Comment by EPD-Analytics** [comment]
Although a workaround (using a project from a home directory) is available, a solution is still desirable. If I have to use my home directory, I am on my OS drive, which is much smaller than my data drive.

**Comment by mizzao** [comment]
Geez, there might be no one commenting on it but that doesn't mean it's not an issue. I guess if you're the dev team you can just ignore issues and they'll get auto-closed!

**Comment by ismail9k** [comment]
It's 2020 already, yet there's no fix for this issue on Windows.

**Comment by WanielDeiss** [comment]
@ismail9k: have you tried my workaround?

it's probably not directly a problem from meteor, but rather a problem of the developer who don't know the meteor workflow in the background.
when you install meteor with the subsystem in a windows filesystem, meteor thinks it is running on a unix machine and install the unix version of the mongodb, this version can't run on a ntfs (windows) filesystem.

**Comment by ismail9k** [comment]
Yes the work around is

**Comment by mizzao** [comment]
> it's probably not directly a problem from meteor, but rather a problem of the developer who don't know the meteor workflow in the background.
> when you install meteor with the subsystem in a windows filesystem, meteor thinks it is running on a unix machine and install the unix version of the mongodb, this version can't run on a ntfs (windows) filesystem.

Isn't the point of WSL that applications can just pretend they're running on Linux?

**Comment by WanielDeiss** [comment]
@mizzao: indeed, you install mongodb for unix. in meteor the mongodb files (bjons) are in your project on a ntfs filesystem. the unix version of mongodb doesn't support a nfts (windows) filesystem. thats it.

i think thats a problem of the ownership differences between the filesystems and that is the reason why it's not directly meteor's problem. mongodb (unix) do not support that.

**Comment by mizzao** [comment]
> @mizzao: indeed, you install mongodb for unix. in meteor the mongodb files (bjons) are in your project on a ntfs filesystem. the unix version of mongodb doesn't support a nfts (windows) filesystem. thats it.

Are you sure? Cause while `/mnt/c/...` is definitely NTFS. `/home/foo/...` has some sort of emulation layer to make it look POSIX compliant, isn't that right?

**Comment by brucejo75** [comment]
I had been meaning to get back to this...

### TL;DR: Just use the workarounds
1. [Symlink](https://github.com/meteor/meteor/issues/9914#issuecomment-405094639), thanks @WanielDeiss!
2. Or I have been running mongo as a Windows service that I can access via my [`MONGO_URL`](https://docs.meteor.com/environment-variables.html#MONGO-URL).

# Some nitty gritty & future tripping

### Why this only works when your project is off the `/home` directory.
It is important to read **[File systems in WSL](https://docs.microsoft.com/en-us/archive/blogs/wsl/wsl-file-system-support#file-systems-in-wsl)** paying special attention to `VolFs`.  Note that `VolFs` is only installed for `/root` & `/home`.

`VolFs` supports enough features for `mongo` to run, yay!

But `VolFs` cannot be used for shared folders with Windows.  those are `DrvFS` file systems that windows understands.  So some features of linux file access are not available on these drives.  And that is what is tripping up mongo.

For WSL, only `VolFs` files can be accessed the way mongo wants to access them.

### But wait maybe there is improvement with WSL 2?
WSL 2 is close to release (maybe 03/2020?), with WSL 2 they take another whack at system file compatibility.

Although WSL 2 has the same restriction 👎  (well actually, I don't know how they could ever get rid of this restriction so it is actually ☹️)

1. They added the capability for [windows access](https://docs.microsoft.com/en-us/windows/wsl/wsl2-ux-changes#place-your-linux-files-in-your-linux-root-file-system) to the root file system. What this means: you can use windows tools to work on files in the root fs (`VolFs` files).  Awesome!
2. Because it is based on a VHD you can easily copy the installation to another location and run it there, e.g your development drive instead of your `C:\` drive.
3. As a bonus, there is significant file system speed up with WSL 2.

I am not scared by the caveats [here](https://docs.microsoft.com/en-us/windows/wsl/wsl2-ux-changes#accessing-network-applications) & [here](https://docs.microsoft.com/en-us/windows/wsl/wsl2-ux-changes#understanding-wsl-2-uses-a-vhd-and-what-to-do-if-you-reach-its-max-size), Microsoft is working to minimize management of your VHD and to make LAN connection more seamless.

I have not tried it yet, I tend to stick to released versions.  But I will jump on this pretty much as soon as it is released, 🤞

**Comment by brucejo75** [comment]
I did verify that `--dbpath` is not an issue.
Also, this is not really a Meteor problem like I first thought it might be.  So I am going to close it.

**Comment by brucejo75** [comment]
PS: I have been using WSL2 for a while now and it is a dream.  I finally relented and started basing my projects under the root filesystem (`VolFs`) and the file speeds are blazing (compared to WSL1 on a windows drive).  And mongo does not seem to be an issue.

Warning: WSL2 `DrvFS` seems slower than WSL1 to the point I found it unbearable...

**Comment by mizzao** [comment]
@brucejo75 amazing, that's great news. What is `DrvFS` though, is that just accessing files on the Windows partition?

**Comment by brucejo75** [comment]
Hi @mizzao, Take a look at my response to you on Jan 7.  Details are in here: https://docs.microsoft.com/en-us/archive/blogs/wsl/wsl-file-system-support#file-systems-in-wsl
